volume/drivers/pwx/connection.go

```diff
@@ -0,0 +1,210 @@
+	"google.golang.org/grpc"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type ConnectionParamsBuilder struct {
+	Config  *ConnectionParamsBuilderConfig
+	kubeOps core.Ops
+}
+
+type ConnectionParamsBuilderConfig struct {
+	// DefaultServiceName is the default name of the Portworx service
+	DefaultServiceName string
+	// DefaultServiceNamespaceName is the kubernetes namespace in which Portworx daemon set runs
+	DefaultServiceNamespaceName string
+	// DefaultRestPortName is name of Porx legacy REST API port in service
+	DefaultRestPortName string
+	// DefaultSDKPortName is name of Porx SDK/iSDK API port in service
+	DefaultSDKPortName string
+
+	// Environment variables names to get  values config properties
+	EnableTLSEnv       string
+	NamespaceNameEnv   string
+	ServiceNameEnv     string
+	CaCertSecretEnv    string
+	CaCertSecretKeyEnv string
+}
+
+var DefaultParams = &ConnectionParamsBuilderConfig{
+	DefaultServiceName:          "portworx-service",
+	DefaultServiceNamespaceName: "kube-system",
+	DefaultRestPortName:         "px-api",
+	DefaultSDKPortName:          "px-sdk",
+	EnableTLSEnv:                "PX_ENABLE_TLS",
+	NamespaceNameEnv:            "PX_NAMESPACE",
+	ServiceNameEnv:              "PX_SERVICE_NAME",
+	CaCertSecretEnv:             "PX_CA_CERT_SECRET",
+	CaCertSecretKeyEnv:          "PX_CA_CERT_SECRET_KEY",
+}
+
+func NewConnectionParamsBuilder(ops core.Ops, params *ConnectionParamsBuilderConfig) *ConnectionParamsBuilder {
+	if params == nil {
+		params = &ConnectionParamsBuilderConfig{}
+	}
+
+	if params.DefaultServiceName == "" {
+		params.DefaultServiceName = DefaultParams.DefaultServiceName
+	}
+
+	if params.DefaultServiceNamespaceName == "" {
+		params.DefaultServiceNamespaceName = DefaultParams.DefaultServiceNamespaceName
+	}
+
+	if params.DefaultRestPortName == "" {
+		params.DefaultRestPortName = DefaultParams.DefaultRestPortName
+	}
+
+	if params.DefaultSDKPortName == "" {
+		params.DefaultSDKPortName = DefaultParams.DefaultSDKPortName
+	}
+
+	if params.EnableTLSEnv == "" {
+		params.EnableTLSEnv = DefaultParams.EnableTLSEnv
+	}
+
+	if params.NamespaceNameEnv == "" {
+		params.NamespaceNameEnv = DefaultParams.NamespaceNameEnv
+	}
+
+	if params.ServiceNameEnv == "" {
+		params.ServiceNameEnv = DefaultParams.ServiceNameEnv
+	}
+
+	if params.CaCertSecretEnv == "" {
+		params.CaCertSecretEnv = DefaultParams.CaCertSecretEnv
+	}
+
+	if params.CaCertSecretKeyEnv == "" {
+		params.CaCertSecretKeyEnv = DefaultParams.CaCertSecretKeyEnv
+	}
+
+	return &ConnectionParamsBuilder{
+		Config:  params,
+		kubeOps: ops,
+	}
+}
+
+// BuildClientsEndpoints returns two endpoints for PX MGMT API and gRPC SDK/iSDL API
+func (cpb *ConnectionParamsBuilder) BuildClientsEndpoints() (pxMgmtEndpoint string, sdkEndpoint string, err error) {
+	var endpoint string
+
+	// Check if service name and namespace is provided
+	// as environment variables
+	serviceName := getPxService(cpb.Config.ServiceNameEnv, cpb.Config.DefaultServiceName)
```

Comment:
nit: These can be part of the object. So that way it would look like this: `cpb.getPxService()` and it always returns the same data for the same object. Same with getPxNamespace()